contracts/common/BasicMetaTransaction.sol

```diff
@@ -0,0 +1,54 @@
+pragma solidity 0.7.3;
+
+import "../../lib/dappsys/math.sol";
+import "./MetaTransactionMsgSender.sol";
+import "./MultiChain.sol";
+
+abstract contract BasicMetaTransaction is DSMath, MetaTransactionMsgSender, MultiChain {
+
+  event MetaTransactionExecuted(address user, address payable relayerAddress, bytes functionSignature);
+
+  function getMetatransactionNonce(address _user) public view virtual returns (uint256 nonce);
+
+  // NB if implementing this functionality in a contract with recovery mode,
+  // you MUST prevent the metatransaction nonces from being editable with recovery mode.
+  function incrementMetatransactionNonce(address _user) internal virtual;
+
+  /// @notice Main function to be called when user wants to execute meta transaction.
+  /// The actual function to be called should be passed as param with name functionSignature
+  /// Here the basic signature recovery is being used. Signature is expected to be generated using
+  /// personal_sign method.
+  /// @param _user Address of user trying to do meta transaction
+  /// @param _payload Function call to make via meta transaction
+  /// @param _sigR R part of the signature
+  /// @param _sigS S part of the signature
+  /// @param _sigV V part of the signature
+  // slither-disable-next-line locked-ether
+  function executeMetaTransaction(address _user, bytes memory _payload,
+      bytes32 _sigR, bytes32 _sigS, uint8 _sigV) public payable returns (bytes memory) {
+
+      require(verify(_user, getMetatransactionNonce(_user), getChainId(), _payload, _sigR, _sigS, _sigV), "metatransaction-signer-signature-mismatch");
+      incrementMetatransactionNonce(_user);
+
+      // Append _user at the end to extract it from calling context
+      (bool success, bytes memory returnData) = address(this).call(abi.encodePacked(_payload, METATRANSACTION_FLAG, _user));
+      require(success, "colony-metatx-function-call-unsuccessful");
+
+      emit MetaTransactionExecuted(_user, msg.sender, _payload);
```

Comment:
Should this be msgSender()? What happens if I make a metatransaction to make a metatransaction...?

Reply:
How would that work? I make a transaction, and then I wrap that transaction in the `executeMetaTransaction` call and then send that out as a metaTransaction? I don't see any reason to disallow it...?

Reply:
Sure, but I'm saying that in that scenario, this `msg.sender` would be wrong?

Reply:
Sure, I see that. The "inner" metatransaction should be credited to the signer, not to the person who paid (who would get credit for the "outer" metatransaction call).

Say A signs the Tx, and then B signs the Metatransaction, which C executes, resulting in a second call to execute (where B should be credited as the sender). In that case, `msg.sender` would return C, which is incorrect. So, let's use `msgSender()`.

Reply:
Although... wouldn't `msg.sender` in the inner metatransaction return `address(this)`? Since the metatransaction contract is executing against itself?

Reply:
It would, yes, which is wrong.